kube-controllers/pkg/controllers/node/ipam.go

```diff
@@ -107,6 +141,9 @@ func NewIPAMController(cfg config.NodeControllerConfig, c client.Interface, cs k
 		nodesByBlock:                make(map[string]string),
 		blocksByNode:                make(map[string]map[string]bool),
 		emptyBlocks:                 make(map[string]string),
+		poolsByBlock:                make(map[string]string),
```

Comment:
I know this isn't really your problem, since the code existed this way before, but the number of complex map structures that we're using (and need to keep in sync!) is growing a little wild. 

This is mostly a code restructuring suggestion, so hopefully not too much effort, but I would like to move this code to a model where we build helper types to make managing / abstracting away this complexity a little easier.

e.g.,

```go
type PoolManager struct {
  poolsByBlock
  blocksByPool
  allPools
}

func (p *pooManager) OnBlockUpdate() {}

func (p *poolManager) OnPoolUpdate() {}
```

I don't think this is required to get this PR merged, so if you think it's too big a lift just say so and we can look at refactoring this whole component a little bit sometime down the line.

Reply:
Happy to give it a go and time-box a refactor on this (assuming you mean just the pool management for now)

Reply:
Yeah, I just meant the pool management stuff for now. 

I think let's get this PR in and then we can do a follow-on if you're up for it, or we can fix it all up later.

Reply:
I had already implemented the pool management refactor by the time you posted your follow-up comment :slightly_smiling_face: 

Take a look and let me know, hopefully I captured what you were thinking. I made it package-local since I think it only has utility to the node controller.

If I missed the mark with what you were thinking, perhaps I can pull it out and we can treat it as separate PR as you suggested so that we don't hold this PR up.